kythe/go/util/tools/vnames/apply.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	"kythe.io/kythe/go/platform/vfs"
+	"kythe.io/kythe/go/util/cmdutil"
+	"kythe.io/kythe/go/util/vnameutil"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/google/subcommands"
+)
+
+type rulesFormat string
+
+const (
+	jsonFormat  rulesFormat = "JSON"
+	protoFormat rulesFormat = "PROTO"
+)
+
+type applyRulesCmd struct {
+	cmdutil.Info
+
+	rulesPath string
+	format    string
+
+	allowEmpty bool
+}
+
+var applyRulesInfo = cmdutil.NewInfo("apply-rules", "apply VName rewrite rules to stdin lines",
+	`Usage: apply-rules --rules <path> [--format proto]`)
+
+func (c *applyRulesCmd) SetFlags(flag *flag.FlagSet) {
+	flag.StringVar(&c.rulesPath, "rules", "", "Path to VName rewrite rules file")
+	flag.StringVar(&c.format, "format", string(jsonFormat), `Path of VName rewrite rules file {"JSON", "PROTO"}`)
+	flag.BoolVar(&c.allowEmpty, "allow_empty", false, "Whether to write an empty VName when no rule matches the input")
+}
+func (c *applyRulesCmd) Execute(ctx context.Context, flag *flag.FlagSet, args ...interface{}) subcommands.ExitStatus {
+	rules, err := rulesFormat(strings.ToUpper(c.format)).readFile(ctx, c.rulesPath)
+	if err != nil {
+		return cmdErrorf("reading %q: %v", c.rulesPath, err)
+	}
+	s := bufio.NewScanner(os.Stdin)
+	m := jsonpb.Marshaler{OrigName: true}
+	for s.Scan() {
+		line := s.Text()
+		v, ok := rules.Apply(line)
+		if ok || c.allowEmpty {
+			if err := m.Marshal(os.Stdout, v); err != nil {
+				return cmdErrorf("writing VName: %v", err)
```

Comment:
any use case to also permit proto (text or binary) output as well?

Reply:
I didn't add them because they're not really well-supported as stream formats.